File: clinic_backend/src/services/medicineService.ts
import prisma from "../config/prisma.js";
import { ApiError } from "../utils/apiError.js";
import { createMedicineInput, updateMedicineInput } from "../validation/medicineSchema.js";
import { Prisma } from "@prisma/client";

export const getAllMedicineService = async () => {
  const medicines = await prisma.medicine.findMany({
    orderBy: { updatedAt: "desc" },
  });

  if (medicines.length === 0) {
    throw new ApiError(404, "Medicine nya Kosong");
  }

  return medicines;
};

export const createMedicineService = async (input: createMedicineInput) => {
  const count = await prisma.medicine.count();
  const code = `MED-${String(count + 1).padStart(3, "0")}`;

  const medicine = await prisma.medicine.create({
    data: {
      name: input.name,
      code: code,
      price: input.price,
      stock: input.stock,
      unit: input.unit,
    },
  });

  return medicine;
};

export const getMedicineByIdService = async (id: number) => {
  const getMedicine = await prisma.medicine.findUnique({
    where: {
      id: id,
    },
  });

  if (!getMedicine) {
    throw new ApiError(404, "Medicine yang di cari ga ada");
  }

  return getMedicine;
};

export const updateMedicineService = async (id: number, input: updateMedicineInput) => {
  await getMedicineByIdService(id);

  const updateMedicine = await prisma.medicine.update({
    where: { id },
    data: input,
  });

  return updateMedicine;
};

export const deleteMedicineService = async (id: number) => {
  await getMedicineByIdService(id);

  // Cek apakah obat pernah diresepkan dalam riwayat konsultasi
  const hasConsultations = await prisma.consultationMedicine.findFirst({
    where: { medicineId: id },
  });

  if (hasConsultations) {
    throw new ApiError(
      400,
      "Obat tidak dapat dihapus karena memiliki riwayat resep / rekam medis pasien"
    );
  }

  // Eksekusi hapus jika tidak ada relasi resep
  return await prisma.medicine.delete({
    where: { id },
  });
};

File: clinic_backend_go/internal/service/medicine_service.go
package service

import (
	"clinic_backend_go/internal/models"
	"clinic_backend_go/internal/repository"
	"fmt"
	"time"
)

type CreateMedicineInput struct {
	Code  string  `json:"code" binding:"required"`
	Name  string  `json:"name" binding:"required"`
	Price float64 `json:"price" binding:"required"`
	Stock int     `json:"stock" binding:"required"`
	Unit  string  `json:"unit" binding:"required"` // tablet, botol, strip
}

type UpdateMedicineInput struct {
	Code  *string  `json:"code"`
	Name  *string  `json:"name"`
	Price *float64 `json:"price"`
	Stock *int     `json:"stock"`
	Unit  *string  `json:"unit"`
}

type MedicineService interface {
	GetAllMedicines() ([]models.Medicine, error)
	GetMedicineByID(id uint) (*models.Medicine, error)
	CreateMedicine(input CreateMedicineInput) (*models.Medicine, error)
	UpdateMedicine(id uint, input UpdateMedicineInput) (*models.Medicine, error)
	DeleteMedicine(id uint) error
}

type medicineService struct {
	medRepo repository.MedicineRepository
}

func NewMedicineService(medRepo repository.MedicineRepository) MedicineService {
	return &medicineService{medRepo: medRepo}
}

func (s *medicineService) GetAllMedicines() ([]models.Medicine, error) {
	return s.medRepo.FindAll()
}

func (s *medicineService) GetMedicineByID(id uint) (*models.Medicine, error) {
	return s.medRepo.FindByID(id)
}

func (s *medicineService) CreateMedicine(input CreateMedicineInput) (*models.Medicine, error) {
	newMed := models.Medicine{
		Code:      input.Code,
		Name:      input.Name,
		Price:     input.Price,
		Stock:     input.Stock,
		Unit:      input.Unit,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	if err := s.medRepo.Create(&newMed); err != nil {
		return nil, err
	}
	return &newMed, nil
}

func (s *medicineService) UpdateMedicine(id uint, input UpdateMedicineInput) (*models.Medicine, error) {
	med, err := s.medRepo.FindByID(id)
	if err != nil {
		return nil, err
	}

	if input.Code != nil {
		med.Code = *input.Code
	}
	if input.Name != nil {
		med.Name = *input.Name
	}
	if input.Price != nil {
		med.Price = *input.Price
	}
	if input.Stock != nil {
		med.Stock = *input.Stock
	}
	if input.Unit != nil {
		med.Unit = *input.Unit
	}
	med.UpdatedAt = time.Now()

	if err := s.medRepo.Update(med); err != nil {
		return nil, err
	}
	return med, nil
}

func (s *medicineService) DeleteMedicine(id uint) error {
	_, err := s.medRepo.FindByID(id)
	if err != nil {
		return err
	}

	isUsed, err := s.medRepo.HasConsultations(id)
	if err != nil {
		return err
	}
	if isUsed {
		return fmt.Errorf("Obat tidak dapat dihapus karena pernah diresepkan ke pasien")
	}

	return s.medRepo.Delete(id)
}
